### PR TITLE
Mr/gradle classes

### PR DESCRIPTION
**Pull request by marschwar**
I fixed the tests and removed the dependencies to `CheckstyleReports` and `AbstractCodeQualityPlugin`.

I will write more tests in the days to come.

Make sure to run ` ./gradlew clean publishToMavenLocal` to create the `1.0.0-GRADLE` version of the cli

### CODE REVIEW

**Comment by marschwar** [comment]
@Mauin As far as I can tell, the check task should now work with a root project and a multi module project. If not specified in the submodule, the properties are drawn from the parents.
Should we try to rebase onto the master?
What else needs to be done?

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt [hunk 1]
```diff
@@ -1,119 +1,88 @@
 package io.gitlab.arturbosch.detekt
 
 import groovy.lang.Closure
-import io.gitlab.arturbosch.detekt.invoke.DetektInvoker
+import io.gitlab.arturbosch.detekt.invoke.*
```
Comment (Mauin) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt]: detekt won't like the star imports.

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt [hunk 2]
```diff
@@ -1,119 +1,88 @@
...
-
-		if (reports.xml.isEnabled) {
-			map += "XML" to reports.xml.destination
-		}
-		if (reports.html.isEnabled) {
-			map += "HTML" to reports.html.destination
-		}
-		return map
-	}
+open class Detekt : DefaultTask(), Reporting<DetektReports> {
```
Comment (Mauin) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt]: `SourceTask` indicates that the task will perform actions on sources, which doesn't sound too wrong. Why the change to `DefaultTask`? https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/SourceTask.html

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt [hunk 3]
```diff
@@ -1,75 +1,59 @@
...
-		get() = configProperty.orNull?.asFile()
-		set(value) = configProperty.set(project.resources.text.fromFile(value))
-
-	fun idea(configuration: Action<in IdeaExtension>) {
-		configuration.execute(ideaExtension)
+open class DetektExtension(project: Project) : CodeQualityExtension() {
+	val defaultReportsDir = project.layout.buildDirectory.get()
+			.dir(ReportingExtension.DEFAULT_REPORTS_DIR_NAME)
+			.dir("detekt").asFile
+	val defaultSourceDirectories = project.files("src/main/java", "src/main/kotlin")
```
Comment (Mauin) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt]: Is it possible to dynamically get the source set paths with `project.sourceSets`?

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt (hunk 2 above)
Comment (marschwar) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt]: IMHO extending SourceTask does not have any benefits unless we decide to actually use the includes/excludes mechanism. But in order to do so we should rework the way we pass source files to detekt as it right now does the filtering internally.

**Comment by marschwar** [comment]
@Mauin do you have any idea how to solve the chicken-and-egg-problem that the we can only build the latest gradle plugin with the latest gradle plugin? Especially now that the version is `1.0.0-GRADLE`. Locally I did:
```
./gradlew clean publishToMavenLocal -x dokka -x detekt
./gradlew build -PwarningsAsErrors=true
```
